docs/providers/provider-protocol-reference.md

```diff
@@ -0,0 +1,243 @@
+    log(format!("Catnip: looking for the closest dispenser to {location:?} now..."));
+```
+
+### `random`
+
+#### Signature
+
+```rust
+pub fn random(len: u32) -> Vec<u8>;
+```
+
+#### Description
+
+Query the hosting runtime for a source of randomness. Note that there is no
+guarantee that the random values you obtain this way to be cryptographically
+secure: it is entirely dependent on the runtime implementation. You can see the
+implementation of randomness for the Fiberplane Daemon in the [Provider Runtime
+crate](https://github.com/fiberplane/fiberplane-rs/blob/6679fbd0f1cbdac7c57422ae699e12bb35bed71b/fiberplane-provider-protocol/fiberplane-provider-runtime/src/spec/mod.rs#L116)
+source code.
+
+For example, you could create a
+[getrandom](https://docs.rs/getrandom/latest/getrandom/macro.register_custom_getrandom.html#writing-a-custom-getrandom-implementation) provider using it
+
+```rust
+// In a helper crate like getrandom_fp_provider
+use getrandom::Error;
+
+pub fn use_fp_bindings(buf: &mut [u8]) -> Result<(), Error> {
+    let vals = random(buf.len());
+    buf.copy_from_slice(&vals);
+    Ok(())
+}
+
+// In your provider crate
+use getrandom_fp_provider::use_fp_bindings;
+use getrandom::register_custom_getrandom;
+
+register_custom_getrandom!(use_fp_bindings);
+```
+
+> Note: this example snippet might prove generic and useful enough to be upstreamed
+  by Fiberplane and available as a public helper crate. Stay tuned!
+
+### `now`
+
+#### Signature
+
+```rust
+pub fn now() -> Timestamp;
+```
+
+#### Description
+
+Returns a timestamp of the current time of execution. The usefulness of this
+timestamp comes from what you want to do with timers, or signature
+creation/verification schemes.
+
+## Exported functions
+
+Exported functions are functions that your provider implements, and that the
+runtime will call. In order to make those exported functions usable in Fiberplane,
+they must be annotated with `#[pdk_export]`.
+
+### `get_supported_query_types`
+
+It is usually safer to use the Plugin Development Kit macros to let the compiler
+generate correct implementations of these functions. A correct `pdk_query_types!`
+invocation will generate this function for you. See the
+[tutorial](doc:create-a-provider) for more information about using the PDK
+
+#### Signature
+
+```rust
+pub async fn get_supported_query_types(config: ProviderConfig) -> Vec<SupportedQueryType>;
+```
+
+#### Description
+
+Returns the query types supported by this provider.
+This function allows Studio to know upfront which formats will be
+supported, and which providers (and their query types) are eligible to
+be selected for certain use cases.
+
+### `invoke2`
+
+It is usually safer to use the Plugin Development Kit macros to let the compiler
+generate correct implementations of these functions. A correct `pdk_query_types!`
+invocation will generate this function for you. See the
+[tutorial](doc:create-a-provider) for more information about using the PDK
```

Comment:
```suggestion
It is advised to use the Plugin Development Kit rather than implementing this function manually.
The [`pdk_query_types!`](https://docs.rs/fiberplane-pdk/latest/fiberplane_pdk/prelude/macro.pdk_query_types.html) macro can be used to generate this function for you.
See the [tutorial](doc:create-a-provider) for more information about using the PDK.
```